site/src/sphinx/advanced-retry.rst

```diff
@@ -1,8 +1,225 @@
+    import com.linecorp.armeria.common.HttpRequest;
+    import com.linecorp.armeria.common.HttpResponse;
+    import com.linecorp.armeria.client.retry.RetryingHttpClient;
+    import com.linecorp.armeria.client.retry.RetryStrategy;
+
+    RetryStrategy strategy = RetryStrategy.onServerErrorStatus();
+    HttpClient client = new ClientBuilder(...)
+                              .decorator(HttpRequest.class, HttpResponse.class,
+                                         RetryingHttpClient.newDecorator(strategy))
+                              .build(HttpClient.class);
+
+    client.execute(...).aggregate().join();
+
+or even simply,
+
+.. code-block:: java
+
+    import com.linecorp.armeria.client.HttpClientBuilder;
+
+    RetryStrategy strategy = RetryStrategy.onServerErrorStatus();
+    HttpClient client = new HttpClientBuilder(...)
+                          .decorator(RetryingHttpClient.newDecorator(strategy))
+                          .build();
+
+    client.execute(...).aggregate().join();
+
+That is it. The client will keep attempting until it succeeds or the number of attempts exceeds the default
+number of max attempts. You can configure the ``defaultMaxAttempts`` when making the decorator using
+``RetryingHttpClient.newDecorator(strategy, defaultMaxAttempts)``. Meanwhile, the ``strategy`` will decide to
+retry depending on the response. In this case, the client retries when it receives ``5xx`` response error.
+
+RetryStrategy
+-------------
+
+You can customize the ``strategy`` by implementing RetryStrategy_.
+
+.. code-block:: java
+
+    import com.linecorp.armeria.client.retry.Backoff;
+    import com.linecorp.armeria.common.HttpStatus;
+
+    new RetryStrategy<HttpRequest, HttpResponse>() {
+        final Backoff backoff = RetryStrategy.defaultBackoff;
+
+        @Override
+        public CompletableFuture<Optional<Backoff>> shouldRetry(HttpRequest request,
+                                                                HttpResponse response) {
+            return response.aggregate().handle((result, cause) -> { // don't do get() or join()!
+                if (cause != null) {
+                    if (cause instanceof ResponseTimeoutException) {
+                        return Optional.of(backoff);
+                    }
+                } else if (result.headers().status() == HttpStatus.CONFLICT) {
+                    return Optional.of(backoff);
+                }
+                return Optional.empty(); // return no backoff not to retry anymore
+            });
+        }
+    };
+
+This will retry when the response's status is ``409``.
+
+.. note::
+
+    We declare a ``backoff`` as a member and return it. RetryingClient_ internally tracks the
+    reference of the ``backoff`` and increases the counter one by one. The counter will be used
+    as the ``numAttemptsSoFar`` to get the next delay in Backoff_. we will take a close look what
+    the Backoff_ is at the bottom of this page.
```

Comment:
Perhaps we should explain 'why' more in detail? Like:

- reference of the backoff -> reference of the returned Backoff_
- and increases the counter ... The counter will be used ... to get the next delay Backoff_. -> , increases the counter that keeps the number of attempts made so far, and resets it to 0 when the Backoff_ returned by the strategy is not same as before. Therefore, it is important to return the same Backoff_ instance unless you decided to change your backoff strategy, because otherwise some Backoff_ implementations such as an exponential backoff, which yields a different delay based on the number of retries, will not work as expected.

Also, how about adding both good and bad example which involves an exponential backoff? It doesn't have to thorough like the example above; we could skip non-essential parts with `...` this time to keep things concise.